File: app/topic-preview.js
(function (root) {
  'use strict';
  const THRESHOLD = 1500;
  const cache = new Map();
  const inflight = new Map();
  const enabled = value => value == null || !['false', '0', 'no', 'off'].includes(String(value).toLowerCase());
  function englishGroups(profile) {
    let groups = profile.constraint_groups;
    if (groups != null) {
      if (!Array.isArray(groups) || !groups.length || groups.some(g => !Array.isArray(g) || !g.length)) throw new Error('英文约束无效');
    } else {
      groups = [(profile.keywords || []).filter(x => typeof x === 'string' || (x && enabled(x.enabled))).map(x => typeof x === 'string' ? x : x.keyword || x.query || x.text || '')];
    }
    if (profile.constraint_groups != null && (groups.length > 3 || groups.some(g => g.length > 8) || groups.reduce((n, g) => n * g.length, 1) > 32)) throw new Error('约束超出预览预算');
    if (profile.constraint_groups == null && groups[0].length > 24) throw new Error('关键词超出预览预算');
    return groups.map(group => {
      const terms = group.map(value => {
        const text = typeof value === 'string' ? value.trim().replace(/\s+/g, ' ') : '';
        if (!/[A-Za-z]/.test(text) || /[^\x20-\x7e]/.test(text) || text.length > 240) throw new Error('需要英文关键词');
        return text;
      });
      if (!terms.length) throw new Error('需要英文关键词');
      return Array.from(new Set(terms));
    });
  }
  function scopeFor(mode, asOf) {
    if (!['90', '365', 'starter'].includes(mode) || !/^\d{4}-\d{2}-\d{2}$/.test(asOf || '')) throw new Error('范围无效');
    const end = new Date(asOf + 'T00:00:00Z');
    if (!Number.isFinite(end.getTime()) || end.toISOString().slice(0, 10) !== asOf) throw new Error('日期无效');
    const start = new Date(end.getTime() - (mode === '90' ? 90 : 365) * 86400000);
    const scope = { mode, arxiv: { start: start.toISOString().slice(0, 10), end_exclusive: asOf } };
    if (mode === 'starter') {
      const year = end.getUTCFullYear() - 2, month = end.getUTCMonth();
      const day = Math.min(end.getUTCDate(), new Date(Date.UTC(year, month + 1, 0)).getUTCDate());
      scope.conference = { start: new Date(Date.UTC(year, month, day)).toISOString().slice(0, 10), end_exclusive: asOf };
    }
    return scope;
  }
  function countFilter(groups, start, end) {
    const quote = value => '"' + value.replace(/\\/g, '\\\\').replace(/"/g, '\\"') + '"';
    return '(' + ['published.gte.' + start, 'published.lt.' + end, ...groups.map(group => 'or(' + group.map(t => 'search_tsv.plfts(english).' + quote(t)).join(',') + ')')].join(',') + ')';
  }
  function publicKey(key) {
    if (/^sb_publishable_[A-Za-z0-9_-]+$/.test(key || '')) return true;
    try {
      let text = key.split('.')[1].replace(/-/g, '+').replace(/_/g, '/');
      text += '='.repeat((4 - text.length % 4) % 4);
      const decoded = typeof atob === 'function' ? atob(text) : Buffer.from(text, 'base64').toString();
      return JSON.parse(decoded).role === 'anon';
    } catch (_) { return false; }
  }
  function backendFor(config) {
    const cfg = config || {};
    const shared = cfg.supabase_shared || {}, legacy = cfg.supabase || {};
    const specific = (cfg.source_backends || {}).arxiv || {};
    const backend = { ...legacy, ...shared, ...specific };
    const key = String(backend.publishable_key || backend.anon_key || '');
    const url = new URL(String(backend.url || ''));
    if (!enabled(backend.enabled) || !publicKey(key) || url.protocol !== 'https:' || url.username || url.password || url.search || url.hash) throw new Error('匿名连接不可用');
    const schema = backend.schema || 'public', table = backend.papers_table || 'arxiv_papers';
    if (![schema, table].every(x => /^[A-Za-z_][A-Za-z0-9_]*$/.test(x))) throw new Error('数据库配置无效');
    let base = url.href.replace(/\/$/, '');
    if (!base.endsWith('/rest/v1')) base += '/rest/v1';
    return { endpoint: base + '/' + table, key, schema };
  }
  async function preview(options) {
    options = options || {};
    const { profile = {}, mode, asOf, conferences = [] } = options || {};
    const result = { status: 'unknown', count: null, threshold: THRESHOLD, scope: null, reason: '范围未确认', coverage: { arxiv: { status: 'unknown', completed_windows: 0 }, conference: { status: mode === 'starter' ? 'unavailable' : 'not_requested', selected: conferences } } };
    let groups, backend;
    const config = options.config || (root.SubscriptionsManager && root.SubscriptionsManager.getDraftConfig ? root.SubscriptionsManager.getDraftConfig() : {});
    try { result.threshold = previewThreshold(options.threshold, config); }
    catch (_) { result.threshold = null; result.reason = '预览阈值必须是1–10000整数；范围未确认'; return result; }
    try {
      groups = englishGroups(profile);
      result.scope = scopeFor(mode, asOf);
      backend = backendFor(config);
    } catch (_) {
      result.reason = '英文研究约束、时间范围或匿名连接不可用；未把未确认结果记为0';
      return result;
    }
    // 不把密钥写入缓存键、输出或日志。缓存只在当前页面内短暂保留。
    const identity = JSON.stringify([backend.endpoint, backend.schema, groups, result.scope, conferences, result.threshold]);
    const hit = cache.get(identity);
    if (hit && Date.now() - hit.time < 30000) return JSON.parse(JSON.stringify(hit.result));
    if (inflight.has(identity)) return JSON.parse(JSON.stringify(await inflight.get(identity)));
    const run = async () => {
      const windows = [], start = Date.parse(result.scope.arxiv.start + 'T00:00:00Z');
      let end = Date.parse(asOf + 'T00:00:00Z');
      while (end > start) {
        const lower = Math.max(start, end - 30 * 86400000);
        windows.push([new Date(lower).toISOString().slice(0, 10), new Date(end).toISOString().slice(0, 10)]);
        end = lower;
      }
      result.coverage.arxiv.total_windows = windows.length;
      let total = 0, completed = 0;
      const deadline = Date.now() + 25000;
      for (const [lower, upper] of windows.slice(0, 16)) {
        const remaining = deadline - Date.now();
        if (remaining <= 0) break;
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), Math.min(6000, remaining));
        try {
          const url = new URL(backend.endpoint);
          url.searchParams.set('select', 'id'); url.searchParams.set('limit', '1');
          url.searchParams.set('and', countFilter(groups, lower, upper));
          const headers = { apikey: backend.key, 'Accept-Profile': backend.schema, Prefer: 'count=exact', Range: '0-0' };
          if (!backend.key.startsWith('sb_publishable_')) headers.Authorization = 'Bearer ' + backend.key;
          const response = await fetch(url.href, { method: 'HEAD', headers, signal: controller.signal, redirect: 'error' });
          if (![200, 206].includes(response.status)) break;
          const match = /^(?:\d+-\d+|\*)\/(\d+)$/.exec((response.headers.get('Content-Range') || '').trim());
          if (!match || !Number.isSafeInteger(Number(match[1]))) break;
          total += Number(match[1]); completed++;
          if (total > result.threshold) break;
        } catch (_) { break; }
        finally { clearTimeout(timer); }
      }
      result.coverage.arxiv.status = completed === windows.length ? 'exact' : completed ? 'partial' : 'unknown';
      result.coverage.arxiv.completed_windows = completed;
      result.count = completed ? total : null;
      if (total > result.threshold) {
        result.status = 'lower_bound'; result.reason = '已计数的不重叠时间片已超过' + result.threshold + '；这是范围规模，不是相关性判断';
      } else if (completed === windows.length && mode !== 'starter') {
        result.status = 'exact'; result.reason = '英文关键词FTS命中数已确认；不等于相关论文数量';
      } else {
        result.status = completed ? 'partial' : 'unknown';
        result.reason = mode === 'starter' ? '仅确认arXiv部分规模，会议总量未确认' : '计数预算耗尽或请求未完成，完整规模未确认';
      }
      if (cache.size >= 64) cache.delete(cache.keys().next().value);
      cache.set(identity, { time: Date.now(), result: JSON.parse(JSON.stringify(result)) });
      return result;
    };
    const promise = run(); inflight.set(identity, promise);
    try { return await promise; } finally { inflight.delete(identity); }
  }
  function previewThreshold(value, config) {
    if (value == null) value = config && config.topic_research && config.topic_research.preview_threshold !== undefined ? config.topic_research.preview_threshold : THRESHOLD;
    if (!['number', 'string'].includes(typeof value) || !/^\d+$/.test(String(value)) || !Number.isInteger(Number(value)) || Number(value) < 1 || Number(value) > 10000) throw new Error('预览阈值无效');
    return Number(value);
  }
  const api = { preview, __test: { englishGroups, scopeFor, countFilter, publicKey, backendFor, previewThreshold, clearCache: () => cache.clear() } };
  root.TopicResearchPreview = api;
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
})(typeof window !== 'undefined' ? window : globalThis);

File: src/topic_preview.py
"""匿名 FTS 范围预览：只计数，不抓正文，不调用模型，不承诺相关性。"""

from __future__ import annotations

import base64
import calendar
from datetime import date, timedelta
import hashlib
import json
import math
import re
import time
from urllib.parse import urlsplit

import requests

THRESHOLD = 1500
MAX_REQUESTS = 16
TOTAL_TIMEOUT = 25
CACHE_SECONDS = 30
_CACHE = {}


def _enabled(value):
    return value is None or str(value).lower() not in {"false", "0", "no", "off"}


def english_groups(profile):
    explicit = profile.get("constraint_groups")
    if explicit is not None:
        if (
            not isinstance(explicit, list)
            or not explicit
            or any(not isinstance(g, list) or not g for g in explicit)
        ):
            raise ValueError("英文约束组无效，无法确认范围")
        raw_groups = explicit
    else:
        terms = []
        for item in profile.get("keywords") or []:
            if isinstance(item, str):
                terms.append(item)
            elif isinstance(item, dict) and _enabled(item.get("enabled")):
                terms.append(
                    item.get("keyword") or item.get("query") or item.get("text") or ""
                )
        raw_groups = [terms]
    if explicit is not None and (
        len(raw_groups) > 3
        or any(len(g) > 8 for g in raw_groups)
        or math.prod(len(g) for g in raw_groups) > 32
    ):
        raise ValueError("英文约束过多，无法在交互预算内确认范围")
    if explicit is None and sum(len(g) for g in raw_groups) > 24:
        raise ValueError("英文约束过多，无法在交互预算内确认范围")
    groups = []
    for raw in raw_groups:
        values = []
        for value in raw:
            text = " ".join(str(value).split()) if isinstance(value, str) else ""
            # 不将中文/混合语句送给英文FTS后错误显示0；不自行翻译或扩词。
            if (
                not re.search(r"[A-Za-z]", text)
                or re.search(r"[^\x20-\x7e]", text)
                or len(text) > 240
            ):
                raise ValueError("缺少可直接计数的英文关键词，请先完善英文研究约束")
            if text not in values:
                values.append(text)
        if not values:
            raise ValueError("缺少可直接计数的英文关键词，请先完善英文研究约束")
        groups.append(values)
    return groups


def scope_for(mode, as_of):
    if str(mode) not in {"90", "365", "starter"}:
        raise ValueError("不支持的预览范围")
    end = date.fromisoformat(str(as_of))
    days = 90 if str(mode) == "90" else 365
    result = {
        "mode": str(mode),
        "arxiv": {
            "start": (end - timedelta(days=days)).isoformat(),
            "end_exclusive": end.isoformat(),
        },
    }
    if mode == "starter":
        start = end.replace(
            year=end.year - 2,
            day=min(end.day, calendar.monthrange(end.year - 2, end.month)[1]),
        )
        result["conference"] = {
            "start": start.isoformat(),
            "end_exclusive": end.isoformat(),
        }
    return result


def count_filter(groups, start, end):
    def quote(term):
        return '"' + term.replace("\\", "\\\\").replace('"', '\\"') + '"'

    clauses = ["published.gte." + start, "published.lt." + end]
    clauses += [
        "or(" + ",".join("search_tsv.plfts(english)." + quote(t) for t in group) + ")"
        for group in groups
    ]
    return "(" + ",".join(clauses) + ")"


def public_key(key):
    if re.fullmatch(r"sb_publishable_[A-Za-z0-9_-]+", key or ""):
        return True
    try:
        part = key.split(".")[1]
        payload = json.loads(base64.urlsafe_b64decode(part + "=" * (-len(part) % 4)))
        return payload.get("role") == "anon"
    except (AttributeError, IndexError, ValueError, TypeError):
        return False


def backend_for(config):
    from source_config import get_source_backend

    backend = get_source_backend(config or {}, "arxiv")
    key = str(backend.get("publishable_key") or backend.get("anon_key") or "")
    raw_url = str(backend.get("url") or "").rstrip("/")
    parsed = urlsplit(raw_url)
    if (
        backend.get("enabled") is False
        or not public_key(key)
        or parsed.scheme != "https"
        or not parsed.netloc
        or parsed.username
        or parsed.query
        or parsed.fragment
    ):
        raise ValueError("缺少可用的匿名论文库连接，范围未确认")
    table, schema = str(backend.get("papers_table") or "arxiv_papers"), str(
        backend.get("schema") or "public"
    )
    if any(
        not re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", item) for item in (table, schema)
    ):
        raise ValueError("论文库配置不合法，范围未确认")
    endpoint = raw_url if raw_url.endswith("/rest/v1") else raw_url + "/rest/v1"
    return endpoint + "/" + table, key, schema


def preview_threshold(value=None, config=None):
    if value is None:
        value = ((config or {}).get("topic_research") or {}).get(
            "preview_threshold", THRESHOLD
        )
    if isinstance(value, float) and math.isfinite(value) and value.is_integer():
        value = int(value)
    if (
        isinstance(value, bool)
        or not isinstance(value, (str, int))
        or not re.fullmatch(r"\d+", str(value))
    ):
        raise ValueError("预览阈值必须是1–10000整数")
    number = int(value)
    if not 1 <= number <= 10000:
        raise ValueError("预览阈值必须是1–10000整数")
    return number


def preview(
    *, profile, mode, as_of, conferences=None, config=None, request=None, threshold=None
):
    scope = None
    coverage = {
        "arxiv": {"status": "unknown", "completed_windows": 0},
        "conference": {
            "status": "unavailable" if mode == "starter" else "not_requested",
            "selected": conferences or [],
        },
    }
    base = {
        "status": "unknown",
        "count": None,
        "threshold": THRESHOLD,
        "scope": scope,
        "coverage": coverage,
        "reason": "范围未确认",
    }
    try:
        threshold = preview_threshold(threshold, config)
        base["threshold"] = threshold
    except (ValueError, TypeError, AttributeError):
        base.update(threshold=None, reason="预览阈值必须是1–10000整数；范围未确认")
        return base
    try:
        groups = english_groups(profile or {})
        scope = scope_for(str(mode), as_of)
        base["scope"] = scope
        endpoint, key, schema = backend_for(config)
    except (ValueError, TypeError, AttributeError):
        base["reason"] = "英文研究约束、时间范围或匿名连接不可用；未把未确认结果记为0"
        return base
    identity = hashlib.sha256(
        json.dumps(
            [endpoint, schema, groups, scope, conferences or [], threshold],
            sort_keys=True,
        ).encode()
    ).hexdigest()
    cached = _CACHE.get(identity)
    if cached and time.monotonic() - cached[0] < CACHE_SECONDS:
        return json.loads(json.dumps(cached[1]))
    end = date.fromisoformat(scope["arxiv"]["end_exclusive"])
    start = date.fromisoformat(scope["arxiv"]["start"])
    windows = []
    while end > start:
        lower = max(start, end - timedelta(days=30))
        windows.append((lower.isoformat(), end.isoformat()))
        end = lower
    coverage["arxiv"]["total_windows"] = len(windows)
    total = completed = 0
    deadline = time.monotonic() + TOTAL_TIMEOUT
    call = request or requests.head
    for lower, upper in windows[:MAX_REQUESTS]:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            break
        headers = {
            "apikey": key,
            "Accept-Profile": schema,
            "Prefer": "count=exact",
            "Range": "0-0",
        }
        if not key.startswith("sb_publishable_"):
            headers["Authorization"] = "Bearer " + key
        try:
            response = call(
                endpoint,
                params={
                    "select": "id",
                    "limit": "1",
                    "and": count_filter(groups, lower, upper),
                },
                headers=headers,
                timeout=min(6, remaining),
                allow_redirects=False,
            )
            if response.status_code not in (200, 206):
                break
            match = re.fullmatch(
                r"(?:\d+-\d+|\*)/(\d+)",
                str(response.headers.get("Content-Range") or "").strip(),
            )
            if not match:
                break
            total += int(match[1])
            completed += 1
            if total > threshold:
                break
        except (requests.RequestException, TimeoutError, ValueError):
            break
    coverage["arxiv"].update(
        status=(
            "exact"
            if completed == len(windows)
            else "partial" if completed else "unknown"
        ),
        completed_windows=completed,
    )
    base["count"] = total if completed else None
    if total > threshold:
        base.update(
            status="lower_bound",
            reason=f"已计数的不重叠时间片已超过{threshold}；这是范围规模，不是相关性判断",
        )
    elif completed == len(windows) and mode != "starter":
        base.update(
            status="exact", reason="英文关键词FTS命中数已确认；不等于相关论文数量"
        )
    else:
        base.update(
            status="partial" if completed else "unknown",
            reason=(
                "仅确认arXiv部分规模，会议总量未确认"
                if mode == "starter"
                else "计数预算耗尽或请求未完成，完整规模未确认"
            ),
        )
    if len(_CACHE) >= 64:
        _CACHE.pop(next(iter(_CACHE)))
    _CACHE[identity] = (time.monotonic(), json.loads(json.dumps(base)))
    return base
